File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.core.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.content.ContextCompat;
import io.reactivex.Flowable;
import io.reactivex.processors.PublishProcessor;
import java.io.File;
import java.util.HashSet;
import java.util.Set;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.kiwix.kiwixmobile.core.CoreApp;
import org.kiwix.kiwixmobile.core.NightModeConfig;

/**
 * Manager for the Default Shared Preferences of the application.
 */

@Singleton
public class SharedPreferenceUtil {
  // Prefs
  public static final String PREF_LANG = "pref_language_chooser";
  public static final String PREF_STORAGE = "pref_select_folder";
  public static final String PREF_WIFI_ONLY = "pref_wifi_only";
  public static final String PREF_KIWIX_MOBILE = "kiwix-mobile";
  public static final String PREF_ZOOM = "pref_zoom_slider";
  public static final String PREF_ZOOM_ENABLED = "pref_zoom_enabled";
  public static final String PREF_SHOW_INTRO = "showIntro";
  private static final String PREF_BACK_TO_TOP = "pref_backtotop";
  private static final String PREF_HIDE_TOOLBAR = "pref_hidetoolbar";
  private static final String PREF_FULLSCREEN = "pref_fullscreen";
  private static final String PREF_NEW_TAB_BACKGROUND = "pref_newtab_background";
  private static final String PREF_STORAGE_TITLE = "pref_selected_title";
  private static final String PREF_EXTERNAL_LINK_POPUP = "pref_external_link_popup";
  private static final String PREF_IS_FIRST_RUN = "isFirstRun";
  private static final String PREF_SHOW_BOOKMARKS_CURRENT_BOOK = "show_bookmarks_current_book";
  private static final String PREF_SHOW_HISTORY_CURRENT_BOOK = "show_history_current_book";
  private static final String PREF_HOSTED_BOOKS = "hosted_books";
  public static final String PREF_NIGHT_MODE = "pref_night_mode";
  private SharedPreferences sharedPreferences;
  private final PublishProcessor<String> prefStorages = PublishProcessor.create();
  private final PublishProcessor<NightModeConfig.Mode> nightModes = PublishProcessor.create();

  @Inject
  public SharedPreferenceUtil(Context context) {
    sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
  }

  public boolean getPrefWifiOnly() {
    return sharedPreferences.getBoolean(PREF_WIFI_ONLY, true);
  }

  public boolean getPrefHideToolbar() {
    return sharedPreferences.getBoolean(PREF_HIDE_TOOLBAR, true);
  }

  public boolean getPrefIsFirstRun() {
    return sharedPreferences.getBoolean(PREF_IS_FIRST_RUN, true);
  }

  public boolean getPrefFullScreen() {
    return sharedPreferences.getBoolean(PREF_FULLSCREEN, false);
  }

  public boolean getPrefBackToTop() {
    return sharedPreferences.getBoolean(PREF_BACK_TO_TOP, false);
  }

  public boolean getPrefZoomEnabled() {
    return sharedPreferences.getBoolean(PREF_ZOOM_ENABLED, false);
  }

  public boolean getPrefNewTabBackground() {
    return sharedPreferences.getBoolean(PREF_NEW_TAB_BACKGROUND, false);
  }

  public boolean getPrefExternalLinkPopup() {
    return sharedPreferences.getBoolean(PREF_EXTERNAL_LINK_POPUP, true);
  }

  public float getPrefZoom() {
    return sharedPreferences.getFloat(PREF_ZOOM, 100.0f);
  }

  public String getPrefLanguage(String defaultLanguage) {
    return sharedPreferences.getString(PREF_LANG, defaultLanguage);
  }

  public String getPrefStorage() {
    String storage = sharedPreferences.getString(PREF_STORAGE, null);
    if (storage == null) {
      final File externalFilesDir =
        ContextCompat.getExternalFilesDirs(CoreApp.getInstance(), null)[0];
      storage = externalFilesDir != null ? externalFilesDir.getPath()
        : CoreApp.getInstance().getFilesDir().getPath(); // workaround for emulators
      putPrefStorage(storage);
    }
    return storage;
  }

  public String getPrefStorageTitle(String defaultTitle) {
    return sharedPreferences.getString(PREF_STORAGE_TITLE, defaultTitle);
  }

  public boolean getPrefFullTextSearch() {
    return false; // Temporarily disable multizim for 2.4
    //return sharedPreferences.getBoolean(PREF_FULL_TEXT_SEARCH, false);
  }

  public void putPrefLanguage(String language) {
    sharedPreferences.edit().putString(PREF_LANG, language).apply();
  }

  public void putPrefIsFirstRun(boolean isFirstRun) {
    sharedPreferences.edit().putBoolean(PREF_IS_FIRST_RUN, isFirstRun).apply();
  }

  public void putPrefWifiOnly(boolean wifiOnly) {
    sharedPreferences.edit().putBoolean(PREF_WIFI_ONLY, wifiOnly).apply();
  }

  public void putPrefStorageTitle(String storageTitle) {
    sharedPreferences.edit().putString(PREF_STORAGE_TITLE, storageTitle).apply();
  }

  public void putPrefStorage(String storage) {
    sharedPreferences.edit().putString(PREF_STORAGE, storage).apply();
    prefStorages.onNext(storage);
  }

  public Flowable<String> getPrefStorages() {
    return prefStorages.startWith(getPrefStorage());
  }

  public void putPrefFullScreen(boolean fullScreen) {
    sharedPreferences.edit().putBoolean(PREF_FULLSCREEN, fullScreen).apply();
  }

  public void putPrefExternalLinkPopup(boolean externalLinkPopup) {
    sharedPreferences.edit().putBoolean(PREF_EXTERNAL_LINK_POPUP, externalLinkPopup).apply();
  }

  public boolean showIntro() {
    return sharedPreferences.getBoolean(PREF_SHOW_INTRO, true);
  }

  public void setIntroShown() {
    sharedPreferences.edit().putBoolean(PREF_SHOW_INTRO, false).apply();
  }

  public boolean getShowHistoryCurrentBook() {
    return sharedPreferences.getBoolean(PREF_SHOW_HISTORY_CURRENT_BOOK, true);
  }

  public void setShowHistoryCurrentBook(boolean prefShowHistoryCurrentBook) {
    sharedPreferences.edit()
      .putBoolean(PREF_SHOW_HISTORY_CURRENT_BOOK, prefShowHistoryCurrentBook)
      .apply();
  }

  public boolean getShowBookmarksCurrentBook() {
    return sharedPreferences.getBoolean(PREF_SHOW_BOOKMARKS_CURRENT_BOOK, true);
  }

  public void setShowBookmarksCurrentBook(boolean prefShowBookmarksFromCurrentBook) {
    sharedPreferences.edit()
      .putBoolean(PREF_SHOW_BOOKMARKS_CURRENT_BOOK, prefShowBookmarksFromCurrentBook)
      .apply();
  }

  @NonNull
  public NightModeConfig.Mode getNightMode() {
    return NightModeConfig.Mode.from(
      Integer.parseInt(
        sharedPreferences.getString(
          PREF_NIGHT_MODE,
          "" + AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
        )
      )
    );
  }

  public Flowable<NightModeConfig.Mode> nightModes() {
    return nightModes.startWith(getNightMode());
  }

  public void updateNightMode() {
    nightModes.offer(getNightMode());
  }

  public Set<String> getHostedBooks() {
    return sharedPreferences.getStringSet(PREF_HOSTED_BOOKS, new HashSet<>());
  }

  public void setHostedBooks(Set<String> hostedBooks) {
    sharedPreferences.edit()
      .putStringSet(PREF_HOSTED_BOOKS, hostedBooks)
      .apply();
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.utils

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import androidx.annotation.ChecksSdkIntAtLeast
import androidx.core.content.edit
import androidx.preference.PreferenceManager
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manager for the Default Shared Preferences of the application.
 */

@Singleton
class SharedPreferenceUtil @Inject constructor(val context: Context) {
  private val sharedPreferences: SharedPreferences =
    PreferenceManager.getDefaultSharedPreferences(context)

  var prefIsTest: Boolean
    get() = sharedPreferences.getBoolean(PREF_IS_TEST, false)
    set(prefIsTest) {
      sharedPreferences.edit { putBoolean(PREF_IS_TEST, prefIsTest) }
    }

  val isPlayStoreBuild: Boolean
    get() = sharedPreferences.getBoolean(IS_PLAY_STORE_BUILD, false)

  fun setIsPlayStoreBuildType(isPlayStoreBuildType: Boolean) {
    sharedPreferences.edit { putBoolean(IS_PLAY_STORE_BUILD, isPlayStoreBuildType) }
  }

  @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.R)
  fun isPlayStoreBuildWithAndroid11OrAbove(): Boolean =
    isPlayStoreBuild && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R

  @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.R)
  fun isNotPlayStoreBuildWithAndroid11OrAbove(): Boolean =
    !isPlayStoreBuild && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R

  companion object {
    // Prefs
    const val PREF_LANG = "pref_language_chooser"
    const val PREF_DEVICE_DEFAULT_LANG = "pref_device_default_language"
    const val PREF_STORAGE = "pref_select_folder"
    const val STORAGE_POSITION = "storage_position"
    const val PREF_WIFI_ONLY = "pref_wifi_only"
    const val PREF_KIWIX_MOBILE = "kiwix-mobile"
    const val PREF_SHOW_INTRO = "showIntro"
    const val PREF_IS_TEST = "is_test"
    const val PREF_SHOW_SHOWCASE = "showShowCase"
    const val PREF_BACK_TO_TOP = "pref_backtotop"
    const val PREF_NEW_TAB_BACKGROUND = "pref_newtab_background"
    const val PREF_EXTERNAL_LINK_POPUP = "pref_external_link_popup"
    const val PREF_SHOW_STORAGE_OPTION = "show_storgae_option"
    const val PREF_IS_FIRST_RUN = "isFirstRun"
    const val PREF_SHOW_BOOKMARKS_ALL_BOOKS = "show_bookmarks_current_book"
    const val PREF_SHOW_HISTORY_ALL_BOOKS = "show_history_current_book"
    const val PREF_SHOW_NOTES_ALL_BOOKS = "show_notes_current_book"
    const val PREF_HOSTED_BOOKS = "hosted_books"
    const val PREF_THEME = "pref_dark_mode"
    const val TEXT_ZOOM = "true_text_zoom"
    const val DEFAULT_ZOOM = 100
    const val PREF_MANAGE_EXTERNAL_FILES = "pref_manage_external_files"
    const val PREF_SHOW_MANAGE_PERMISSION_DIALOG_ON_REFRESH = "pref_show_manage_external_files"
    const val IS_PLAY_STORE_BUILD = "is_play_store_build"
    const val PREF_BOOKMARKS_MIGRATED = "pref_bookmarks_migrated"
    const val PREF_RECENT_SEARCH_MIGRATED = "pref_recent_search_migrated"
    const val PREF_HISTORY_MIGRATED = "pref_history_migrated"
    const val PREF_NOTES_MIGRATED = "pref_notes_migrated"
    const val PREF_APP_DIRECTORY_TO_PUBLIC_MIGRATED = "pref_app_directory_to_public_migrated"
    const val PREF_BOOK_ON_DISK_MIGRATED = "pref_book_on_disk_migrated"
    const val PREF_SHOW_COPY_MOVE_STORAGE_SELECTION_DIALOG = "pref_show_copy_move_storage_dialog"
    const val PREF_LATER_CLICKED_MILLIS = "pref_later_clicked_millis"
    const val PREF_LAST_DONATION_POPUP_SHOWN_IN_MILLISECONDS =
      "pref_last_donation_shown_in_milliseconds"
    const val SELECTED_ONLINE_CONTENT_LANGUAGE = "selectedOnlineContentLanguage"
    const val CACHED_LANGUAGE_CODES = "cachedLanguageCodes"
    const val KEY_LANGUAGE_CODE = "languageCode"
    const val KEY_OCCURRENCES_OF_LANGUAGE = "occurrencesOfLanguage"
    const val KEY_LANGUAGE_ACTIVE = "languageActive"
    const val KEY_LANGUAGE_ID = "languageId"
    const val PREF_SCAN_FILE_SYSTEM_DIALOG_SHOWN = "prefScanFileSystemDialogShown"
    const val PREF_IS_SCAN_FILE_SYSTEM_TEST = "prefIsScanFileSystemTest"
  }
}
